File: config-shared-core-kt/assets/shared-template-kt/src/main/kotlin/com/example/shared/domain/vo/Name.kt
package com.example.shared.domain.vo

import com.example.shared.domain.result.DomainResult

@JvmInline
value class Name private constructor(val value: String) {
    companion object {
        private const val INVALID_NAME = "Name must not be blank and must have at most 255 characters"

        fun create(value: String): Name =
            tryCreate(value).getOrThrow()

        fun tryCreate(value: String): DomainResult<Name> {
            val errors = mutableListOf<String>()
            val normalized = value.trim()
            if (normalized.isBlank()) errors.add("Name must not be blank")
            if (normalized.length > 255) errors.add("Name must have at most 255 characters")
            if (errors.isNotEmpty()) return DomainResult.failure(errors)
            return DomainResult.success(Name(normalized))
        }
    }
}

File: config-shared-core-cs/assets/shared-template-cs/Domain/ValueObjects/Name.cs
using Project.Shared.Kernel.Domain.Results;

namespace Project.Shared.Kernel.Domain.ValueObjects;

public record Name
{
    public string Value { get; }

    private Name(string value) => Value = value;

    public static Result<Name> Create(string value)
    {
        if (string.IsNullOrWhiteSpace(value))
            return Result<Name>.Failure("Name is required");

        if (value.Length < 3)
            return Result<Name>.Failure("Name must be at least 3 characters");

        return Result<Name>.Success(new Name(value.Trim()));
    }

    public static implicit operator string(Name name) => name.Value;
}
